interfaces/prompting/patterns/render.go

```diff
@@ -0,0 +1,376 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patterns
+
+import (
+	"bytes"
+	"strings"
+)
+
+// RenderConfig is a configuration of a render node.
+type RenderConfig interface {
+	// NextEx modifies the configuration to the next state, if any remain.
+	// Returns the length in bytes of the new configuration, the number of
+	// bytes which remain unchanged since the previous configuration, and true
+	// if more configurations remain to be rendered. The argument is always the
+	// render node that was earlier used to obtain render configuration.
+	NextEx(n RenderNode) (length, lengthUnchanged int, moreRemain bool)
```

Comment:
I really don't know why this is called NextEx? what does Ex stands for? why do we need to pass in the config again if it's the one we got the config to start with?

Reply:
Ex stands for Expansion, but I agree `NextVariant` would be much clearer, along with `variantState` as you suggest above.

We need to pass in the node because the configs don't hold references to the configs themselves, just configs (now `variantState`s). I'm not sure why we don't embed them, perhaps @zyga can provide a rationale.

Reply:
I increasingly think `Render` should be a method of `variantState` rather than `renderNode`, and `variantState` should hold its affiliated `renderNode`.

Reply:
I've made this change. Now there is no need to pass a `renderNode` or `variantState` as a parameter to any method call, and `Render` is a method of `variantState`.